### PR TITLE
Adds virology tumors that spread and (maybe) kill you

### PR DESCRIPTION
**Pull request by ToasterBiome**
# Document the changes in your pull request

Adds tumors to virology. 

I am aware that this bounty is not approved yet (nor do I have any control if it will be approved or not)
https://forums.yogstation.net/threads/30-1-4-add-tumors-as-a-virology-symptom-benign-premalignant-malignant.24324/

Benign, Premalignant, and Malignant tumors.
Benign, Premalignant, and Malignant virology symptoms.

if you have a disease with the tumor symptom, you will have a chance to get a tumor organ in a random slot in your body. Using a medical scanner will tell you where the tumor is.
if you have tumors, you have a chance to have them spread to other areas of your body.
if you remove all the tumors associated with a disease, you get cured.
Benign tumors are mostly harmless, the higher versions are stronger and spread faster.

Transmission 7 makes it so you have visible growths on your body, these will progressively make it so you can wear less and less things depending on stages: stage 3 you lose the ability to wear hats/masks. stage 5 you cant wear chest slot items.
Stealth 4 makes it so you have a chance to regenerate limbs
Resistance 8 makes you heal brute/burn damage for each tumor you have. 

Benign Tumors:
0.125 toxic damage per life tick
if sufficient resistance, heal 0.0625 brute and 0.0625 burn per life tick
0.25% chance to spread per life tick

Premalignant Tumors:
0.25 toxic damage per life tick
if sufficient resistance, heal 0.125 brute and 0.125 burn per life tick
0.5% chance to spread per life tick

Malignant Tumors:
0.5 toxic damage per life tick
if sufficient resistance, heal 0.25 brute and 0.25 burn per life tick
1% chance to spread per life tick

I have no clue if this is balanced and I need feedback

need to still be implemented: 

- [x] Tumor Sprites (just uses heart sprite for now)
- [x] Nicer tumor names (just says the slot, so right arm is "r_arm")
- [x] Limb regeneration 
- [x] visible growths locking you out of clothing slots

# Wiki Documentation

3 new virology symptoms, 3 new organs, something about removing them all curing a disease

# Changelog

:cl:  
rscadd: Adds 3 new tumors
rscadd: Adds 3 new virology symptoms
tweak: lets organs be inserted into special funny slots
/:cl:

### CODE REVIEW

**Comment by Mqiib** [comment]
So, assuming I'm reading this right, you'll get exponential growth of tumors because each tumor has its own spread proc. At 0.06 tumors per tumor per two seconds we can get a spicy equation set up for how many tumors (spoilers: it's a lot) you can get over the course of your time left (not very long).
image
The red line is the number of tumors, green line is total toxin damage taken by the tumors, x is seconds.
As you can see, about 88 seconds after the first tumor (roughly since we're dealing with single tumors and not fractions of tumors) you'll be in crit from toxin damage AND have about 12 more buddies with you. At around the 109 second mark, you're dead from toxin damage and have 24 or so friends. If you're NOT dead at this point you're taking an exponentially increasing 12 toxin damage per second (that's six times what charcoal heals). If you somehow survive for a whole two minutes, you'll have 33 friends to be surgically removed dealing a combined 16.5 damage per cycle.
But what about the baby tumors? Certainly they're fine?
image
Not really. Instead of being in crit at 88 seconds, you'll stay up for another 200 seconds as you slowly begin to realize what's going on. Death still follows shortly afterward at the roughly 355 second mark, about 70 seconds later. At this point you'll have around 31 friends coming to your funeral.

This needs some limits to how many tumors you can have and a SIGNIFICANT decrease in how fast they grow.

**Comment by ToasterBiome** [comment]
you can only have 8 tumors

you only get 8 body parts where tumors can grow, and tumors wont grow in places that already have tumors in them

**Comment by Mqiib** [comment]
That still leaves only about 70 seconds to get all 8 malignant tumors and 4 toxin damage per cycle perpetually until you get EIGHT surgeries done to fix it.

**Comment by ToasterBiome** [comment]
> That still leaves only about 70 seconds to get all 8 malignant tumors and 4 toxin damage per cycle perpetually until you get EIGHT surgeries done to fix it.

what would it look like if all the spread times were divided by 2? 4? 8?

**Comment by Mqiib** [comment]
> > That still leaves only about 70 seconds to get all 8 malignant tumors and 4 toxin damage per cycle perpetually until you get EIGHT surgeries done to fix it.
> 
> what would it look like if all the spread times were divided by 2? 4? 8?

half growth speed = double time needed to hit max

**Comment by ToasterBiome** [comment]
> > > That still leaves only about 70 seconds to get all 8 malignant tumors and 4 toxin damage per cycle perpetually until you get EIGHT surgeries done to fix it.
> > 
> > 
> > what would it look like if all the spread times were divided by 2? 4? 8?
> 
> half growth speed = double time needed to hit max

oh also if it picks that zone to put a tumor into and you already have a tumor it just... doesnt put one in that life tick. does that change the calculations

**Comment by Mqiib** [comment]
yes that would make it complicated enough that i dont want to go through them again but it would also effectively reduce the growth speed

**Comment by ToasterBiome** [comment]
@adamsong sanity review plz????

**Comment by ToasterBiome** [comment]
@adamsong remove do not merge tag please it is ready

**Review by adamsong — APPROVED** [review]
Seems to look good code wise, balance is a different concern though

**Comment by SomeguyManperson** [comment]
what exactly is the purpose of this symptom is it supposed to be helpful or harmful? Symptoms like RPS which can kill you if not treated properly feel like noob bait and just get shoved into healing viruses because they "look cool" without any way to actually manage them, tumors would require you to constantly have access to some kind of toxin healing chem or other system to be useful or it'll kill you which makes it seem more wack to me than stuff like plasma fixation which require access to x chemical but WON'T kill you if you don't have it
also use add_no_equip_slot and remove_no_equip_slot procs to reduce the amount of bullshit that might show up from pre-existing no_equip flags

**Comment by ToasterBiome** [comment]
> what exactly is the purpose of this symptom is it supposed to be helpful or harmful? Symptoms like RPS which can kill you if not treated properly feel like noob bait and just get shoved into healing viruses because they "look cool" without any way to actually manage them, tumors would require you to constantly have access to some kind of toxin healing chem or other system to be useful or it'll kill you which makes it seem more wack to me than stuff like plasma fixation which require access to x chemical but WON'T kill you if you don't have it also use add_no_equip_slot and remove_no_equip_slot procs to reduce the amount of bullshit that might show up from pre-existing no_equip flags

threshholded RPS deals the same amount of toxin damage in the same amount of time as four of the worst tier of tumors. I think they're supposed to be harmful, but if you get the threshholds get some benefits like the healing, and limb regeneration. So.. harmful until it's helpful?

**Comment by ToasterBiome** [comment]
I halved the toxin damage if it's thresholded to be "helpful", so now at max it's _only_ as dangerous as threshholded RPS

File: code/datums/diseases/advance/symptoms/tumors.dm [hunk 1]
```diff
@@ -82,6 +75,9 @@
 		if(missing_limbs.len > 0)
 			var/limb_to_regenerate = pick(missing_limbs)
 			M.regenerate_limb(limb_to_regenerate,TRUE);
+			var/obj/item/bodypart/new_limb = M.get_bodypart(limb_to_regenerate);
```
Comment (ToasterBiome) [inline code/datums/diseases/advance/symptoms/tumors.dm]: is this the right way to make a really fragile limb? or does brute + burn have to add up to < 50 as max_damage is 50?

File: code/datums/diseases/advance/symptoms/tumors.dm [hunk 2]
```diff
@@ -44,22 +44,15 @@
...
-				M.dropItemToGround(wearing_mask)
-		if(istype(M.wear_mask, /obj/item/clothing/head))
-			var/obj/item/clothing/head/wearing_hat = M.head
-			if(M.canUnEquip(wearing_hat))
-				M.dropItemToGround(wearing_hat)
-		M.dna.species.no_equip.Add(SLOT_WEAR_MASK,SLOT_HEAD)
+		var/datum/species/S = M.dna?.species
+		if(S)
+			S.add_no_equip_slot(M, SLOT_WEAR_MASK)
+			S.add_no_equip_slot(M, SLOT_HEAD)
```
Comment (ToasterBiome) [inline code/datums/diseases/advance/symptoms/tumors.dm]: @SomeguyManperson is this the correct way to use add_no_equip_slot?

**Comment by ToasterBiome** [comment]
tumor spread chances have been buffed after some testmerge testing, now twice as strong. If thresholded properly you only take toxin damage when you're above brute + burn of when exchanging brut + burn damage for toxins will actually help you.

**Review by nesisknesis — APPROVED** [review]

**Comment by ToasterBiome** [comment]
@SomeguyManperson More feedback on this would be nice to get this to a state you're happy with, it's been testmerged for a month with no issues.

File: code/datums/diseases/advance/symptoms/tumors.dm [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+		regeneration = TRUE
+	if(A.totalResistance() >= 8) //helpful healing instead of just toxin
+		helpful = TRUE
+
+/datum/symptom/tumor/Activate(datum/disease/advance/A)
+	. = ..()
+	if(!.)
+		return
+	var/mob/living/carbon/human/M = A.affected_mob
+	if(!M) return
```
Comment (Bibby0110) [inline code/datums/diseases/advance/symptoms/tumors.dm]: seperate

File: code/modules/surgery/organs/tumors.dm [hunk 4]
```diff
@@ -0,0 +1,80 @@
+/obj/item/organ/tumor
+	name = "benign tumor"
+	desc = "Hope there aren't more of these."
+	icon_state = "tumor"
+
+	var/strength = 0.125
```
Comment (Bibby0110) [inline code/modules/surgery/organs/tumors.dm]: comment + maybe define

for both

File: code/modules/surgery/organs/tumors.dm [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	if(!(src in owner.internal_organs))
+		Remove(owner)
+	if(helpful)
+		if(owner.getBruteLoss() + owner.getFireLoss() > 0)
+			owner.adjustToxLoss(strength/2)
+			owner.adjustBruteLoss(-(strength/2))
+			owner.adjustFireLoss(-(strength/2))
+	else
+		owner.adjustToxLoss(strength) //just take toxin damage
+		//regerenation
```
Comment (Bibby0110) [inline code/modules/surgery/organs/tumors.dm]: .

File: code/modules/surgery/organs/tumors.dm [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+	if(regeneration && prob(spread_chance))
+		var/list/missing_limbs = owner.get_missing_limbs() - list(BODY_ZONE_HEAD, BODY_ZONE_CHEST) //don't regenerate the head or chest
+		if(missing_limbs.len > 0)
```
Comment (Bibby0110) [inline code/modules/surgery/organs/tumors.dm]:
```suggestion
		if(missing_limbs.len)
```

File: code/modules/surgery/organs/tumors.dm [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+			var/limb_to_regenerate = pick(missing_limbs)
+			owner.regenerate_limb(limb_to_regenerate,TRUE);
```
Comment (Bibby0110) [inline code/modules/surgery/organs/tumors.dm]:
```suggestion
			owner.regenerate_limb(limb_to_regenerate,TRUE)
```

File: code/modules/surgery/organs/tumors.dm [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+			var/obj/item/bodypart/new_limb = owner.get_bodypart(limb_to_regenerate);
```
Comment (Bibby0110) [inline code/modules/surgery/organs/tumors.dm]:
```suggestion
			var/obj/item/bodypart/new_limb = owner.get_bodypart(limb_to_regenerate)
```

File: code/datums/diseases/advance/symptoms/tumors.dm [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+				S.add_no_equip_slot(M, SLOT_WEAR_MASK)
+				S.add_no_equip_slot(M, SLOT_HEAD)
+
+		if(A.stage == 5)
+			var/datum/species/S = M.dna?.species
+			if(S)
+				S.add_no_equip_slot(M, SLOT_WEAR_SUIT)
+
+	//spreading
+	if(prob(tumor_chance)) //2% chance to make a new tumor somewhere
```
Comment (Bibby0110) [inline code/datums/diseases/advance/symptoms/tumors.dm:61]: please don't have this repeated in two places.

**Review by Bibby0110 — CHANGES_REQUESTED** [review]

File: code/modules/surgery/organs/tumors.dm (hunk 5 above)
Comment (ToasterBiome) [reply code/modules/surgery/organs/tumors.dm]: what could this mean

File: code/datums/diseases/advance/symptoms/tumors.dm (hunk 9 above)
Comment (ToasterBiome) [reply code/datums/diseases/advance/symptoms/tumors.dm:61]: where do you suggest it goes

File: code/modules/surgery/organs/tumors.dm (hunk 4 above)
Comment (ToasterBiome) [reply code/modules/surgery/organs/tumors.dm]: both what?

File: code/datums/diseases/advance/symptoms/tumors.dm (hunk 3 above)
Comment (ToasterBiome) [reply code/datums/diseases/advance/symptoms/tumors.dm]: seperate what

**Comment by JamieD1** [comment]
Please Resolve your conversations/suggestions when you have done them

**Comment by ToasterBiome** [comment]
> Please Resolve your conversations/suggestions when you have done them

@TheGamerdk please reply to my responses so I may move this PR forward

File: code/datums/diseases/advance/symptoms/tumors.dm (hunk 9 above)
Comment (Bibby0110) [reply code/datums/diseases/advance/symptoms/tumors.dm:61]: i don't care, the ugly way if you cant find a way to properly OOP (which i don't think we can) would just be a global proc or something in this file

**Comment by Bibby0110** [comment]
> > Please Resolve your conversations/suggestions when you have done them
> 
> @TheGamerdk please reply to my responses so I may move this PR forward



dont resolve unresolved stuff pls

**Comment by Bibby0110** [comment]
oh you actually did it, just ignore me

**Review by Bibby0110 — APPROVED** [review]

**Pull request merged**